core/src/main/java/org/elasticsearch/rest/action/RestActions.java

```diff
@@ -244,6 +245,9 @@ private static QueryBuilder parseTopLevelQueryBuilder(XContentParser parser) {
         try {
             QueryBuilder queryBuilder = null;
             for (XContentParser.Token token = parser.nextToken(); token != XContentParser.Token.END_OBJECT; token = parser.nextToken()) {
+                if (token == null) {
+                    throw new EOFException("Unexpected end of JSON request body.");
```

Comment:
I'd rather like to have a check outside the parsing loop that asserts that the first token the parser emmits is XContentParser.Token.START_OBJECT. I think its save to assume we are expecting full json objects. I'd also just throw a ParsingException in that case.

Reply:
@cbuescher but if we just verify that the first token is` XContentParser.Token.START_OBJECT`, we'll still see this infinite loop for `"{"` for example. Do you mean we want to accept that?
Is it really a good idea to have the behavior of `org.elasticsearch.rest.action.RestActions#parseTopLevelQueryBuilder` be to loop forever on part of a valid JSON request? :)

Reply:
> we'll still see this infinite loop for "{" for example.

I'd expect the parser to throw an exception on this?

> Is it really a good idea to have the behavior of org.elasticsearch.rest.action.RestActions#parseTopLevelQueryBuilder be to loop forever on part of a valid JSON request? :)

Of course not, and this is not what @cbuescher said.

Reply:
@tlrx

> I'd expect the parser to throw an exception on this?

Currently, it doesn't  throw on half a valid JSON request that's where my confusion was coming from. Also just goes into the infinite loop then.

Reply:
I just implemented the tests @cbuescher asked for below (good call obviously it started throwing on empty string now too :)).
The best fix I see that includes the empty string case as well would be to do this:

```java
            if (parser.nextToken() == null) {
                return null;
            }
            for (XContentParser.Token token = parser.nextToken(); token != XContentParser.Token.END_OBJECT; token = parser.nextToken()) {
                if (token == null) {
                    throw new ParsingException(parser.getTokenLocation(), "Failed to parse");
                }
```

if the first token isn't a `XContentParser.Token.START_OBJECT` the parser throws eventually anyways and we handle the empty string case without much code, how about it? :)

Reply:
updated the PR with the above suggestion

Reply:
That won't loop forever, as far as I can see this will run into a JsonEOFException because if the unmatched brackets eventually.

Reply:
@cbuescher I tried it out by hand:

```sh
curl -XGET  -H 'Content-Type: application/json' http://localhost:9200/\*/_count -d '"{"'
```

blocks forever

Reply:
Worse yet, it prevents ES from exiting even after I kill `curl`

Reply:
Thats a String token though...

Reply:
Wops my bad :) you're right, '{' dies just fine.
So basically we have `return null` if the first token is `null` (to cover the empty string case), throw if it's not null and not `ContentParser.Token.START_OBJECT` and leave the loop untouched? :)

Reply:
++

Reply:
And because this is really a bit confusing (I was suprised by some of this myself), maybe add some of the above cases in tests. Or wdyt @tlrx

Reply:
@cbuescher I agree

Reply:
@cbuescher @tlrx done :) Added all the discussed cases here https://github.com/elastic/elasticsearch/pull/26680/files#diff-2098a5df6345871e4c3fdb934d3f9f49R72